provisioning/baremetal_pod.go

```diff
@@ -468,10 +476,73 @@ func NewMetal3Deployment(targetNamespace string, images *Images, config *metal3i
 	}
 }
 
-func GetExistingMetal3Deployment(client appsclientv1.DeploymentsGetter, targetNamespace string) (bool, error) {
-	existing, err := client.Deployments(targetNamespace).Get(context.Background(), baremetalDeploymentName, metav1.GetOptions{})
+func annotateExistingMetal3Deployment(existing *appsv1.Deployment) {
+	curAnnotations := existing.GetAnnotations()
+	if _, ok := curAnnotations[cboOwnedAnnotation]; ok {
+		// metal3 deployment already owned by CBO
+		return
+	}
+	cboAnnotations := map[string]string{}
+	cboAnnotations[cboOwnedAnnotation] = ""
+	modified := resourcemerge.BoolPtr(true)
+	resourcemerge.MergeMap(modified, &cboAnnotations, existing.GetAnnotations())
+	existing.SetAnnotations(cboAnnotations)
+	return
+}
+
+func ApplyNewOrUpdatedDeployment(client appsclientv1.DeploymentsGetter, targetNamespace string, metal3Deployment *appsv1.Deployment, generations []osoperatorv1.GenerationStatus, componentName string) error {
+	finalDeployment := metal3Deployment
+
+	// Check if there is an existing metal3 deployment
+	existing, err := client.Deployments(targetNamespace).Get(context.Background(), BaremetalDeploymentName, metav1.GetOptions{})
+
 	if existing != nil && err == nil {
-		return true, nil
+		if equality.Semantic.DeepEqual(metal3Deployment.Spec, existing.Spec) {
+			// the deployment spec of the existing deployment already matches what is required
+			// So, may need to just update annotation and generation
+
+			annotateExistingMetal3Deployment(existing)
+			finalDeployment = existing
+		}
+	}
+	expectedGeneration := resourcemerge.ExpectedDeploymentGeneration(finalDeployment, generations)
+	deployment, updated, err := resourceapply.ApplyDeployment(client,
+		events.NewLoggingEventRecorder(componentName), finalDeployment, expectedGeneration)
+	if err != nil {
+		return errors.Wrap(err, "failed to apply new metal3 Deployment")
+	}
+	if updated {
+		resourcemerge.SetDeploymentGeneration(&generations, deployment)
```

Comment:
generations is not passed to this function as a pointer, so any update here will be lost when this function returns.

Reply:
Fixed this.